File: python/leetcode/array/945_min_inc_unique.py
"""
945. Minimum Increment to Make Array Unique (Medium)

Given an array of integers A, a move consists of choosing any A[i], and incrementing it by 1.

Return the least number of moves to make every value in A unique. 

Example 1:

Input: [1,2,2]
Output: 1
Explanation:  After 1 move, the array could be [1, 2, 3].
Example 2:

Input: [3,2,1,2,1,7]
Output: 6
Explanation:  After 6 moves, the array could be [3, 4, 1, 2, 5, 7].
It can be shown with 5 or less moves that it is impossible for the array to have all unique values.

Note:

0 <= A.length <= 40000
0 <= A[i] < 40000
"""

class Solution(object):
    def minIncrementForUnique(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        A.sort()
        # tmp_max = max(A)
        result = 0
        memo = []
        for i in range(1, len(A)):
            if A[i] == A[i-1]:
                memo.append(A[i])
            elif A[i] - A[i-1] > 1 and len(memo) > 0:
                for j in range(A[i-1]+1, A[i]):
                    if len(memo) > 0:
                        item = memo.pop()
                        result += j - item
                    else:
                        break
        if len(memo) > 0:
            tmp_max = max(A)
            for item in memo:
                result += tmp_max + 1 - item
                tmp_max += 1
        return result
    
    # smarter! 
    def minIncrementForUnique(self, A):
        res = need = 0
        for i in sorted(A):
            res += max(need - i, 0)
            need = max(need + 1, i + 1)
        return res 

File: cc/leetcode/array/945_min_inc_unique.cc
class Solution {
public:
    int minIncrementForUnique(vector<int>& A) {
        sort(A.begin(), A.end());
        A.push_back(INT_MAX);
        vector<int> tmp;
        int result = 0;
        for (int i = 1; i < A.size(); ++i) {
            if (A[i] == A[i-1])
                tmp.push_back(A[i]);
            if (A[i] - A[i-1] > 1) {
                for (int j = A[i-1]+1; j < A[i]; ++j) {
                    if (tmp.size() > 0) {
                        result += j - tmp.back();
                        tmp.pop_back();
                    } else
                        break;
                }
            }
        }
        return result;
    }
};

// Smarter!
int minIncrementForUnique(vector<int>& A) {
        sort(A.begin(), A.end());
        int res = 0, need = 0;
        for (int a: A) {
            res += max(need - a, 0);
            need = max(a, need)+1;
        }
        return res;
    }
